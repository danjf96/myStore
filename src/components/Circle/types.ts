import { TextStyle, ViewStyle } from "react-native";

export interface CIRCLEPROPS {
    style?: ViewStyle,
    styleText?: TextStyle,
    number?: number
}