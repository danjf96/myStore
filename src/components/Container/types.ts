import React from "react";
import { ViewProps, ViewStyle } from "react-native";

export interface PROPSCONTAINER extends ViewProps {
    style?: ViewStyle
}