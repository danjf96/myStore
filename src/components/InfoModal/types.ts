import { ImageRequireSource, ModalProps } from "react-native";

export interface PROPSMODALINFO extends ModalProps {
    image?: ImageRequireSource
    title: string,
    text: string,
    textButton?: string,
    onPress: () => void
    textSecondButton?: string,
    onPressSecondButton?: ()=> void
}
