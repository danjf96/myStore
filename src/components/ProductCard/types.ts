import { PRODUCTSTATE } from "../../store/ducks/products/types";

export interface ProductsCardProps extends PRODUCTSTATE {
    number?: number,
    quantity?: boolean,
    onPressQuantity?: (type: 'increase' | 'decrease') => void,
    testId?: string
}