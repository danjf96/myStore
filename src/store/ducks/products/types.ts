export interface STATESTORE {
    loading: boolean,
    listProducts: PRODUCTSTATE[]
}

//TYPES ACTIONS
export const Types = {
    CLEAN: 'CLEAN',
    CHANGE: 'CHANGE_PRODUCT',
    LOAD: 'LOAD_PRODUCT'
}

export interface PRODUCTSTATE {
    id: string,
    createdAt: string,
    name: string,
    price: number,
    image: string,
    stock: number
}
