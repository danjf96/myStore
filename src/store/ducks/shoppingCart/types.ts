export interface STATESTORE {
    loading: boolean,
    cart: CARTSTATE[],
    totalPrice: number,
    payment?: 'creditCard' | 'balance'
}

//TYPES ACTIONS
export const Types = {
    CLEAN: 'CLEAN',
    CLEAN_CART: 'CLEAN_CART',
    CLEAN_SHOPPINGCART: 'CLEAN_SHOPPINGCART',
    CHANGE: 'CHANGE_SHOPPINGCART',
    CHANGE_VALUE: 'CHANGE_SHOPPINGCART',
    LOAD: 'LOAD_SHOPPINGCART'
}

export interface CARTSTATE {
    id: string,
    createdAt: string,
    name: string,
    price: number,
    image: string,
    stock: number,
    number: number
}