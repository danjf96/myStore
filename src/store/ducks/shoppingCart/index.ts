import api, { errorHandling } from "../../../services/api";
import { STATESTORE, Types } from "./types";


//REDUCERS
const INITIAL_STATE: STATESTORE = {
    loading: false,
    cart: []
}

//STATE CASES
const STATE = (state = INITIAL_STATE, action: any):STATESTORE => {
    switch (action.type) {
        case Types.LOAD:
            return { ...state, loading: action.loading }
        case Types.CHANGE:
            return { ...state, ...action.payload }
        case Types.CHANGE_VALUE:
            return { ...state, [action.key]: action.value }
        case Types.CLEAN:
            return INITIAL_STATE
        default:
            return state;
    }
}

//Actions Creators
export const changeShoppingCart = (payload:{}) => {
    return async (dispatch: (arg0:any) => any) => {
        dispatch({ type: Types.CHANGE, payload  })
    }
}

export const changeShoppingCartValue = (key:string, value:any) => {
    return async (dispatch: (arg0:any) => any) => {
        dispatch({ type: Types.CHANGE, key, value  })
    }
}
export default STATE