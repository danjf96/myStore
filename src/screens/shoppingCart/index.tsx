import React, { useEffect, useState } from 'react'
import { FlatList, Text, TouchableOpacity, View } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import { GlobalStyle } from '../../assets/GlobalStyles'
import Container from '../../components/Container'
import InfoModal from '../../components/InfoModal'
import ProductCard from '../../components/ProductCard'
import { changeShoppingCart } from '../../store/ducks/shoppingCart'
import { formatMoney } from '../../utils/mask'
import Styles from './styles'

const ShoppingCart = (props:any) => {

    const dispatch = useDispatch()
    const { shoppingCart: { cart, loading, totalPrice } } = useSelector( ({ shoppingCart  } ) => ({  shoppingCart }) )
    const [ openModal, setOpenModal ] = useState(true)

    const changeQuantity = (type: string, id:string) => {
        let newCart = cart
        switch(type){
            case 'increase':
                newCart = newCart.map( n => n.id == id ? { ...n, number: n.number += 1 } : n )
            break;
            case 'decrease':
                if(newCart.find( n => n.id == id && n.number - 1 == 0))
                    newCart = newCart.filter( n => n.id != id)
                else
                    newCart = newCart.map( n => n.id == id ? { ...n, number: n.number -= 1 } : n )
            break
        }
        const totalPrice = newCart.length > 0 ? newCart.map( n => n.price * n.number).reduce( (a,b) => a+b) : 0
        dispatch( changeShoppingCart({ cart: newCart, totalPrice }))

    }

    return (
        <>
            <Container style={{ height: '100%', paddingBottom: 120}}>
                <Text style={Styles.price}> Total: R$ {formatMoney(totalPrice)}</Text>

                <View style={GlobalStyle.line}></View>
                
                <FlatList 
                    renderItem={ ({ item, index}) => <ProductCard 
                            {...item}
                            testId={'cart'}
                            quantity={true}
                            styleContainer={Styles.productContainer}
                            onPressQuantity={ t => changeQuantity(t, item.id)}
                        />
                    }
                    keyExtractor={ (item,index) => `cart${index}`}
                    data={cart}
                />   

                <InfoModal  
                    testID='finishModal'
                    visible={openModal}
                    title={'Finalizar compra!'} 
                    text={'Escolha uma forma de pagamento'} 
                    textButton={'USAR CARTÃƒO CADASTRADO'} 
                    textSecondButton='USAR SALDO DISPONIVEL' 
                    onPress={() => null} 
                    onPressSecondButton={() => null}
                    onClose={ () => setOpenModal(false)}
                />  

            </Container>

            <View style={Styles.footer}>
                <TouchableOpacity style={Styles.buttonTwo} onPress={ () => props.navigation.navigate('Home')} >
                    <Text style={Styles.buttonTextTwo}>CONTINUAR COMPRANDO</Text>
                </TouchableOpacity>

                <TouchableOpacity style={Styles.button} onPress={ () => setOpenModal(true)} >
                    <Text style={Styles.buttonText}>FINALIZAR COMPRA</Text>
                </TouchableOpacity>
            </View>
            
        </>

    )
}

export default ShoppingCart