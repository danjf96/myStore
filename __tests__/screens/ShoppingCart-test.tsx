/**
 * @format
 */

 import 'react-native';
 import React from 'react';
 import ShoppingCart from '../../src/screens/shoppingCart';
 import { fireEvent, render, waitFor } from 'react-native-testing-library';
import { Provider } from 'react-redux';
import store from '../../src/store';
import { CARTSTATE, Types } from '../../src/store/ducks/shoppingCart/types';
  
 const component = (
    <Provider store={store} > <ShoppingCart navigation={{ navigate: (screen:'string') => null }}/> </Provider>
  );
 it('renders correctly', async () => {
   const screen = render(component)
   expect(screen.toJSON()).toMatchSnapshot()
 });
  
describe('more tests', () => {
    const storeCart = store.getState().shoppingCart
    const fakeValues:CARTSTATE[] = [{ 
        createdAt: "2019-09-02T12:58:54.103Z",
         id: "1",
         image: "http://lorempixel.com/640/480/food",
         name: "Rustic Metal Fish",
         price: 289.00,
         stock: 65171,
         number: 1
     }]

    test('list empty', () => {   
        expect(storeCart.cart).toHaveLength(0)
    })

    test('list', async () => {   
        const screen = render(component)
        store.dispatch({ type: Types.CHANGE, payload: { cart: fakeValues }})
        await waitFor( () => {
            const cart = store.getState().shoppingCart.cart
            expect(cart).not.toHaveLength(0)
            expect(screen.getAllByTestId('cart')).not.toHaveLength(0)
        })
    })

    test('change quantity', async () => {   
        const { getAllByTestId, getByText } = render(component)
        store.dispatch({ type: Types.CHANGE, payload: { cart: fakeValues }})
        const card = getAllByTestId('cart')[0]
        fireEvent( getByText('+') , 'press')
        expect(getByText('2')).toBeTruthy()

        fireEvent( getByText('-') , 'press')
        expect(getByText('1')).toBeTruthy()
    })

})