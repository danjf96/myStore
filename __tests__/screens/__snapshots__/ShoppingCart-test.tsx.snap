// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  " ",
  <RCTSafeAreaView
    emulateUnlessSupported={true}
  >
    <View
      style={
        Object {
          "height": "100%",
          "padding": 15,
          "paddingBottom": 120,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#0086ff",
            "fontSize": 25,
            "fontWeight": "bold",
            "marginTop": 10,
          }
        }
      >
         Total: R$ 
        0,00
      </Text>
      <View
        style={
          Object {
            "borderColor": "#0086ff",
            "borderWidth": 0.3,
            "marginVertical": 25,
            "opacity": 0.2,
            "width": "100%",
          }
        }
      />
      <RCTScrollView
        data={Array []}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        viewabilityConfigCallbackPairs={Array []}
      >
        <View />
      </RCTScrollView>
      <Modal
        animationType="fade"
        hardwareAccelerated={false}
        onClose={[Function]}
        onPress={[Function]}
        onPressSecondButton={[Function]}
        testID="finishModal"
        text="Escolha uma forma de pagamento"
        textButton="USAR CARTÃO CADASTRADO"
        textSecondButton="USAR SALDO DISPONIVEL"
        title="Finalizar compra!"
        transparent={true}
        visible={false}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#353535",
                "flexDirection": "column",
                "height": "100%",
                "justifyContent": "center",
                "opacity": 0.8,
                "position": "absolute",
                "width": "100%",
              }
            }
          />
          <View
            style={
              Object {
                "paddingHorizontal": 40,
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "white",
                  "borderRadius": 10,
                  "padding": 24,
                  "paddingTop": 32,
                  "width": "100%",
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                nativeID="animatedComponent"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                    "position": "absolute",
                    "right": 10,
                    "top": 10,
                    "zIndex": 99,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "#0086ff",
                        "fontSize": 25,
                      },
                      undefined,
                      Object {
                        "fontFamily": "anticon",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  
                </Text>
              </View>
              <Text
                style={
                  Object {
                    "color": "black",
                    "fontSize": 20,
                    "fontWeight": "bold",
                    "lineHeight": 19.6,
                    "marginBottom": 24,
                  }
                }
              >
                Finalizar compra!
              </Text>
              <Text
                style={
                  Object {
                    "color": "black",
                    "fontSize": 16,
                    "letterSpacing": 0.15,
                    "lineHeight": 24,
                    "marginBottom": 24,
                    "textAlign": "left",
                  }
                }
              >
                Escolha uma forma de pagamento
              </Text>
              <View
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                <View
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#0086ff",
                      "borderRadius": 10,
                      "height": 56,
                      "justifyContent": "center",
                      "marginBottom": 15,
                      "opacity": 1,
                      "padding": 10,
                      "width": "100%",
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "white",
                        "fontSize": 16,
                        "fontWeight": "bold",
                      }
                    }
                  >
                    USAR CARTÃO CADASTRADO
                  </Text>
                </View>
                <View
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "white",
                      "borderColor": "#0086ff",
                      "borderRadius": 10,
                      "borderWidth": 1.5,
                      "height": 56,
                      "justifyContent": "center",
                      "opacity": 1,
                      "padding": 10,
                      "width": "100%",
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#0086ff",
                        "fontSize": 16,
                        "fontWeight": "bold",
                      }
                    }
                  >
                    USAR SALDO DISPONIVEL
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </Modal>
    </View>
  </RCTSafeAreaView>,
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "white",
        "bottom": 0,
        "flexDirection": "column",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderColor": "#0086ff",
          "borderRadius": 10,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "marginBottom": 10,
          "opacity": 1,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#0086ff",
            "fontSize": 18,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        CONTINUAR COMPRANDO
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0086ff",
          "borderRadius": 10,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "marginBottom": 10,
          "opacity": 1,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        FINALIZAR COMPRA
      </Text>
    </View>
  </View>,
  " ",
]
`;
